[Metric Monitoring] Add Kubernetes Support

simon-mo opened the pull request:
This PR adds metrics monitoring to Kubernetes container manager. 

Todos:
- [x] Integration Test
- [x] Doc
- [x] ~~Modify the k8s example~~  Just add a readme

simon-mo committed:
Allow port specification for Prometheus

simon-mo committed:
Allow port specification for Prometheus

simon-mo committed:
Add monitoring example with grafana

This commit adds an example in the example folder. It helps
the user to visualize clipper metric.

The init_grafana.py script launches a grafana/grafana docker
container. It adds prometheus as a data source. I attempted
to add the dashboard via Grafana API but failed after trying
it out for hours. This seems like a persistent issue for at
least two years: https://github.com/grafana/grafana/issues/2816

<Simon Mo>

simon-mo committed:
Add monitoring example with grafana

This commit adds an example in the example folder. It helps
the user to visualize clipper metric.

The init_grafana.py script launches a grafana/grafana docker
container. It adds prometheus as a data source. I attempted
to add the dashboard via Grafana API but failed after trying
it out for hours. This seems like a persistent issue for at
least two years: https://github.com/grafana/grafana/issues/2816

<Simon Mo>

simon-mo committed:
Fix dict unpacking lint error

simon-mo committed:
Format Code

simon-mo committed:
Update yapf

simon-mo committed:
Update yapf and reformat code

simon-mo committed:
Check yapf version in shell script

simon-mo committed:
Change `echo` to `cat`

simon-mo committed:
Add python version check; almost ready

simon-mo committed:
Format code with python version 2.7.12

simon-mo committed:
Update Metric Config; Finish Implement Model Container

simon-mo committed:
Add Integration Test

simon-mo committed:
Fix Integration Tests

simon-mo committed:
Format Code and Rebase from Develop

simon-mo committed:
Merge branch 'metrics' of https://github.com/simon-mo/clipper into metrics

simon-mo committed:
Format Code Again with python2.7.12

simon-mo committed:
Merge branch 'develop' into metrics

simon-mo committed:
Update monitoring readme; trigger Jenkins

simon-mo committed:
Merge branch 'metrics' of https://github.com/simon-mo/clipper into metrics

simon-mo committed:
Address Comments

1. Fix naming and typos
2. Update configuration buckets

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into metrics

simon-mo committed:
Format Code

simon-mo committed:
Address comments

- use 1000.0 instead 1000 rpc arithmatic.
- change docs

dcrankshaw committed:
Merge branch 'develop' into metrics

simon-mo committed:
Address RPC Test issue

Add a block of code make sure config file is found.

simon-mo committed:
Merge branch 'metrics' of https://github.com/simon-mo/clipper into metrics

simon-mo committed:
Merge branch 'develop' into metrics

simon-mo committed:
Update tensowflowcifar

simon-mo committed:
Merge branch 'metrics' of https://github.com/simon-mo/clipper into metrics

simon-mo committed:
[Metric] Update metrics example showing anomaly

simon-mo committed:
[Metric] Fix version tag for prometheus

simon-mo committed:
[Metric] Change version tag to master

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into metrics

simon-mo committed:
Format Code

withsmilo committed:
Support latest conda version

* Fix issue #382

dcrankshaw committed:
Merge branch 'develop' into support_latest_conda_ver

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into metrics

simon-mo committed:
Refactor monitoring example

simon-mo committed:
Merge branch 'support_latest_conda_ver' of https://github.com/withsmilo/clipper into kube_metrics

simon-mo committed:
Add example script for docker and k8s

simon-mo committed:
Add prometheus related config yml

simon-mo committed:
Modify query-frontend yaml

simon-mo committed:
Update k8s container manager

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into kube_metrics

simon-mo committed:
Format Code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1015/
Test FAILed.

simon-mo committed:
Fix lint with py2.7.14

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1016/
Test FAILed.

simon-mo commented:
The cause of previous integration test failure is that query-frontend restarted several times. The restart is due to failure in `frontend-exporter`. 
Since `frontend-exporter` is too fragile and fail often, I'm separating it to a separate deployment (no need to service because prometheus directly get pod's metric)

simon-mo committed:
Separate frontend-exporter as depolyment outside query-frontend

1. Move it to deployment
2. Because it's no longer in the same pod as query-frontend, we
now need to directly call against query-frontend service.
3. Update fronte_end_exporter image to accept ports.
4. Update docker metric for corresponding change

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1018/
Test FAILed.

simon-mo committed:
Format Code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1019/
Test FAILed.

simon-mo committed:
Take off the unnecessary node port

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1020/
Test FAILed.

simon-mo commented:
@dcrankshaw The kubernetes test pass locally when I used my own docker account as container registry. 

The error message seems to be about container crash loop back off with `frontend-exporter`. I changed the frontend-exporter.py and expect the image to be rebuilt in jenkins. Is it possible the kuberntes test is not pulling the right image? Because I used `clipper/frontend-exporter` in the deployment file (https://github.com/ucbrise/clipper/pull/411/files#diff-a2c1103f5133e9ac391d38ed0fd7331dR19), I don't know if the testing cluster is pulling the one just created by jenkins or the one in dockerhub. And I don't know how to find out. Can you help?

simon-mo committed:
Add version to frontend-exporter

simon-mo committed:
Format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1021/
Test FAILed.

simon-mo commented:
@dcrankshaw I don’t understand what went wrong. The image is built but Docker says not found:
```
Successfully tagged clipper/frontend-exporter:fa91a4d

real	0m1.558s
user	0m0.444s
sys	0m0.452s
The push refers to a repository [568959175238.dkr.ecr.us-west-1.amazonaws.com/clipper/query_frontend]
f795babc2e87: Preparing
51e19c6ff330: Preparing
2c7d29cd337a: Preparing
f3a093b335e7: Preparing
2f5b0990636a: Preparing
c9748fbf541d: Preparing
b3968bc26fbd: Preparing
aa4e47c45116: Preparing
788ce2310e2f: Preparing
b3968bc26fbd: Waiting
788ce2310e2f: Waiting
aa4e47c45116: Waiting
f3a093b335e7: Layer already exists
2c7d29cd337a: Layer already exists
2f5b0990636a: Layer already exists
aa4e47c45116: Layer already exists
b3968bc26fbd: Layer already exists
c9748fbf541d: Layer already exists
788ce2310e2f: Layer already exists
f795babc2e87: Pushed
51e19c6ff330: Pushed
develop: digest: sha256:5da63f5334615877e6f92ea3621f1b46db4cded4498e1249d6164b08634d2ecf size: 2207
The push refers to a repository [568959175238.dkr.ecr.us-west-1.amazonaws.com/clipper/management_frontend]
2625bb4f7630: Preparing
51e19c6ff330: Preparing
2c7d29cd337a: Preparing
f3a093b335e7: Preparing
2f5b0990636a: Preparing
c9748fbf541d: Preparing
b3968bc26fbd: Preparing
aa4e47c45116: Preparing
788ce2310e2f: Preparing
b3968bc26fbd: Waiting
aa4e47c45116: Waiting
788ce2310e2f: Waiting
c9748fbf541d: Waiting
2f5b0990636a: Layer already exists
2c7d29cd337a: Layer already exists
f3a093b335e7: Layer already exists
aa4e47c45116: Layer already exists
c9748fbf541d: Layer already exists
b3968bc26fbd: Layer already exists
788ce2310e2f: Layer already exists
2625bb4f7630: Pushed
51e19c6ff330: Pushed
develop: digest: sha256:5692ee79a80a06eb776fffa5c6030b1382e1c967f233891ddc7f0e02a21519f9 size: 2207
The push refers to a repository [568959175238.dkr.ecr.us-west-1.amazonaws.com/clipper/frontend-exporter]
73ff25926f53: Preparing
c7d7c964e43d: Preparing
b9bd82fc1962: Preparing
6dce5c484bde: Preparing
057c34df1f1a: Preparing
3d358bf2f209: Preparing
0870b36b7599: Preparing
8fe6d5dcea45: Preparing
06b8d020c11b: Preparing
b9914afd042f: Preparing
4bcdffd70da2: Preparing
b9914afd042f: Waiting
8fe6d5dcea45: Waiting
06b8d020c11b: Waiting
4bcdffd70da2: Waiting
0870b36b7599: Waiting
3d358bf2f209: Waiting
name unknown: The repository with name 'clipper/frontend-exporter' does not exist in the registry with id '568959175238'
Build step 'Execute shell' marked build as failure
Test FAILed.
```

simon-mo commented:
~~Waiting on #413~~

dcrankshaw commented:
jenkins retest this please

dcrankshaw commented:
The docker error is a configuration issue on my end. Just fixed it.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1025/
Test PASSed.

simon-mo committed:
Few clean up

1. Add port specification. Make scrape targets constraint to
those having prometheus.io/scrape annotation.
2. Remove k8s example because too similar to docker
3. Add get_metric_port function in k8s manager

simon-mo committed:
Move metrics integration test specific dir

simon-mo committed:
Add integration test for k8s_metrics

simon-mo committed:
Update doc

simon-mo committed:
Format Code

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into kube_metrics

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1028/
Test FAILed.

simon-mo committed:
Fix path in metric test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1029/
Test FAILed.

simon-mo committed:
Print result, sleep longer, pred fewer

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1030/
Test FAILed.

simon-mo commented:
Seems like pyspark issue.

simon-mo commented:
Jenkins test this please

simon-mo commented:
jenkins retest this please

simon-mo committed:
update readme + retrigger the test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1033/
Test FAILed.

dcrankshaw commented on clipper_admin/clipper_admin/docker/docker_metric_utils.py:
```diff
@@ -46,6 +46,7 @@ def run_query_frontend_metric_image(name, docker_client, query_name,
 
     query_frontend_metric_cmd = "--query_frontend_name {}:{}".format(
         query_name, CLIPPER_INTERNAL_QUERY_PORT)
+
```
revert

simon-mo replied:
Fixed.

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -157,6 +170,19 @@ def connect(self):
+            metrics_ports = self._k8s_v1.read_namespaced_service(
+                name="metrics", namespace='default').spec.ports
+            for p in metrics_ports:
+                if p.name == "9090":
```
Where does 9090 come from?

simon-mo replied:
The prometheus port.

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -157,6 +170,19 @@ def connect(self):
+
+            query_addr = "{}:{}".format(self.external_node_hosts[0],
+                                        self.clipper_query_port)
+            start_frontend_exporter(self._k8s_beta, query_addr)
```
Why are you starting the frontend exporter in `connect()`? Shouldn't it only be started in `start_clipper`?

simon-mo replied:
Fixed.

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -111,35 +114,44 @@ def start_clipper(self, query_frontend_image, mgmt_frontend_image,
                     body=body, namespace='default')
+
+        start_prometheus(self._k8s_v1, self._k8s_beta)
+
```
Shouldn't you start the `frontend_exporter` here?

simon-mo replied:
Make sense. I just need to get the query_addr again.

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -183,6 +209,10 @@ def deploy_model(self, name, version, input_type, image, num_replicas=1):
+                            },
+                            'annotations': {
+                                "prometheus.io/scrape": "true",
+                                "prometheus.io/port": "1390"
```
Are these Clipper-specific, or is this something prometheus will automatically pick up?

simon-mo replied:
This is something prometheus will automatically pick up. I documented them here: https://github.com/simon-mo/clipper/tree/kube_metrics/monitoring#scrape-target-discovery

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_metric_utils.py:
```diff
@@ -0,0 +1,81 @@
+        _k8s_v1.create_namespaced_config_map(body=data, namespace='default')
+
+
+def _create_prometheus_depolyment(_k8s_beta):
```
typo in deployment

simon-mo replied:
Fixed

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_metric_utils.py:
```diff
@@ -0,0 +1,81 @@
+
+
+def _create_prometheus_depolyment(_k8s_beta):
+    with open(prom_depolyment_path, 'r') as f:
```
type in deployment here too

simon-mo replied:
Fixed

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_metric_utils.py:
```diff
@@ -0,0 +1,81 @@
+        _k8s_v1.create_namespaced_service(body=data, namespace='default')
+
+
+def _create_frontend_exporter_depolyment(_k8s_beta, query_addr):
```
typo in deployment

simon-mo replied:
Fixed

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_metric_utils.py:
```diff
@@ -0,0 +1,81 @@
+        _k8s_v1.create_namespaced_service(body=data, namespace='default')
+
+
+def _create_frontend_exporter_depolyment(_k8s_beta, query_addr):
```
What was the reasoning behind the decision to have the frontend exporter in a separate deployment from the query frontend, rather than in the same pod?

simon-mo replied:
My initial concern is that: If they are in the same pod, the failure of frontend-exporter will cause continuous restart within the pod. Query-frontend is critical and frontend-exporter is not. 

But if we put them in the same pod, it will definitely make our lives easier when we implement query-frontend scale out. 

I just put the exporter back into the same pod

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/prom_configmap.yml:
```diff
@@ -0,0 +1,35 @@
+apiVersion: v1
```
What is the configmap for?

simon-mo replied:
configMap stores Prometheus configuration data in Kubernetes. Kubernetes then pass the configuration to the prometheus pod when the pod is spawned.

simon-mo committed:
Use noop instead of python

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/prom_depolyment.yml:
```diff
@@ -0,0 +1,37 @@
+        volumeMounts:
+        - name: config-volume #This name needs to match the one under "volumes"
+          mountPath: /etc/prometheus
+      volumes:
```
What is the volume for?

simon-mo replied:
This is a volume to configMap, it overrides the default prometheus configuration

dcrankshaw commented on examples/monitoring/README.md:
```diff
@@ -4,3 +4,7 @@ Steps:
+
+For Kubernetes:
+
+- You can find the metric address via `KubernetesContainerManager::get_metric_addr` function. 
```
nit: change to "\`KubernetesContainerManager.get_metric_addr\`"

simon-mo replied:
Fixed

dcrankshaw commented on examples/monitoring/docker/query.py:
```diff
@@ -62,7 +62,12 @@ def fizz_buzz(i):
+        "doubles",
+        feature_sum,
+        num_replicas=2,
+        base_image="simonmok/py-rpc")
```
Presumably this should be updated to `clipper/py-rpc`?

simon-mo replied:
Fixed.

dcrankshaw commented on integration-tests/metrics/k8s_metric.py:
```diff
@@ -0,0 +1,110 @@
+    logger = logging.getLogger(__name__)
+
+    logger.info("Start K8s Metric Test (0/1): Running 2 Replicas")
+    clipper_conn = clipper_conn = create_kubernetes_connection(
```
You have a double `clipper_conn = ` here

simon-mo replied:
Fixed.

dcrankshaw reviewed (changes requested):
This looks great. My comments are all pretty minor.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1034/
Test FAILed.

simon-mo committed:
Address Comments

simon-mo committed:
Rename file

simon-mo committed:
Format code

simon-mo committed:
Python nameError

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1035/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1036/
Test FAILed.

simon-mo committed:
Fix imagePullBackOff issue

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1037/
Test FAILed.

simon-mo committed:
Sleep longer

simon-mo committed:
Always pull exporter image to prevent caching

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1038/
Test FAILed.

simon-mo commented:
This one needs restart. I don’t think it’s due to my changes.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1043/
Test FAILed.

simon-mo committed:
Metric test make more pred + k8 test less bursty pred

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1052/
Test FAILed.

simon-mo commented:
```
18-02-26:04:09:56 INFO     [clipper_admin.py:388] Building model Docker image with model data from data
Traceback (most recent call last):
  File "../integration-tests/metrics/k8s_metric.py", line 81, in <module>
    "568959175238.dkr.ecr.us-west-1.amazonaws.com/clipper")
  File "/clipper/clipper_admin/clipper_admin/clipper_admin.py", line 298, in build_and_deploy_model
ec2-54-187-245-87.us-west-2.compute.amazonaws.com:32334
    container_registry)
  File "/clipper/clipper_admin/clipper_admin/clipper_admin.py", line 390, in build_model
    fileobj=context_file, custom_context=True, tag=image)
  File "/opt/conda/lib/python2.7/site-packages/docker/models/images.py", line 183, in build
    raise BuildError(chunk['error'], result_stream)
docker.errors.BuildError: failed to register layer: open /var/lib/docker/aufs/layers/782d04a3bc9ae768c8c02b9a7af153a79f2190659f7ea186c464660f909a6474: no such file or directory
```
what happened to you jenkins 😟

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1053/
Test FAILed.

dcrankshaw commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1058/
Test FAILed.

dcrankshaw commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1059/
Test FAILed.

dcrankshaw commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1060/
Test FAILed.

simon-mo commented:
Closed due to #427
